_docs/rfc/dynamo_table_config.md

```diff
@@ -0,0 +1,47 @@
+# Allow further customization of Terraform Lock table for S3 Remote State
+
+**STATUS**: In proposal
+
+## Background
+
+When Terragrunt creates a DynamoDB table, it sets the throughput to the default,
+`PROVISIONED`. Provisioned is good for stable throughput, but Terraform largely
+works with variable workloads, often going unused.
+
+AWS also offers a `PAY_PER_REQUEST` billing mode that charges based on the
+number of requests to the lock table. The pricing is dirt cheap, and it will
+further reduce the costs associated with using Terragrunt to manage Terraform's
+base requirements.
+
+Terraform has solved this problem: It involves creating and managing the
+resource manually. But, if Terragrunt is going to be managing the table, it
+should allow for additional configuration to be layered on top of it via
+Terragrunt.
+
+## Proposed solution
+
+### `dynamodb_table_config` block in `remote_state.config`
+
+Because our company has a need for it, I've implemented a solution in my own
+fork that accomplishes this.
+
+https://github.com/ThatGerber/terragrunt/tree/add-dynamodb-billing-mode
+
+The `dynamodb_table_config` block accepts several arguments that are forwarded
+through to the DynamoTableInput object, pulling in additional values from the
+parent config as defaults.
+
+To receive the full values, I had to write a few interfaces for the methods
+being called (like `GetLockTableName`) and change the function calls to accept
+the interfaces instead of strings.
+
+### `dynamodb_table_config` block that populates `dynamodb.CreateTableInput` struct
+
+Another option is to create a config attribute that accepts any arguments that
+would be sent to `CreateTable` or `UpdateTable` map the values through to the
+methods when they're being applied.
```

Comment:
While this works, the drawbacks of this approach is that it commits us to keeping this remote block up to date with the AWS API, which means that we will always be "chasing" updates in the SDK. This duplicates logic available with other tools.

This also sets a precedent that opens us up for doing the same with a bunch of other resources (e.g S3, GCS buckets) and I am not sure we want to commit to maintaining an API like that.

The ideal solution I would be looking for for a feature like this would be to switch the terragrunt routines to use terraform under the hood. It would be really nice if you could write out the terraform resource block for S3 buckets and DynamoDB tables somewhere (this could be what you have for `dynamodb_table_config`), and then terragrunt directly forwards that to the terraform AWS provider to manage the state backend resources. It would be even better if we could store the state somewhere so that you could update and destroy the resources in the end as well.

For the short term though, I think we should just default the creation routine to `PAY_PER_REQUEST` instead of `PROVISIONED`.

Reply:
I understand the concern. I want to say, "But AWS rarely changes the APIs in backwards incompatible ways so it won't be a big deal", but it becomes a bigger dependency and that's always going to be difficult to deal with. 

I do like the idea of reading/writing from state, and could write up something in a few days to to read the provider/resource schema to generate the resource, but as you said, it needs to be written out to a file within the repo that Terraform can read (sans -grunt) otherwise running `terraform -` won't see it. And then either way, -grunt picks up a new dependency.

If it's agreed to update the default billing mode, I'll create a new PR for now to update create table call to set it to `PAY_PER_REQUEST`, and open an additional RFC to discuss how to manage buckets and lock tables going forward?

Reply:
Branch here: https://github.com/ThatGerber/terragrunt/tree/dynamodb-table-billing-mode

Reply:
Do we know that `PAY_PER_REQUEST` is available in all partitions/regions? The announcement said it was available in all commercial regions, but what about `aws-us-gov`, or `aws-cn`? Or the new partitions exposed in recent SDKs, `us-iso` and `us-isob`?

Reply:
Not sure about ISO, but it is available in govcloud (https://aws.amazon.com/about-aws/whats-new/2019/06/amazon-dynamodb-on-demand-and-transactions-now-available-in-the-aws-govcloud-us-regions/) and CN (https://www.amazonaws.cn/en/dynamodb/pricing/).

With that said, that is a good point. Does anyone with access to ISO partitions know if it is available and if the error message from AWS is clear if it is not available? Perhaps we can catch that error and fall back to `PROVISIONED` if the first call with `PAY_PER_REQUEST` fails?

Alternatively, we can make this configurable... although in that case we would basically be introducing `dynamodb_table_config`.

Reply:
DynamoDB is available in `iso`/`iso-b` https://github.com/aws/aws-sdk-go/blob/master/aws/endpoints/defaults.go#L5278 and on the gov-cloud announcement page it says:

> Support for on-demand and transactions is available in all AWS Regions where DynamoDB is available.

Reply:
the `iso` partitions are not govcloud. gimme a bit please, i am trying to get a straight answer...

Reply:
Is the accepted solution then that the default provisioning should be switched to `PAY_PER_REQUEST`? Would you like me to submit a new RFC proposing this or should I submit a merge request?

Reply:
Thanks for confirming that @lorengordon !

@ThatGerber At this point you can actually open a PR with the change and keep this open for further discussion to enable additional configuration.